core/src/main/scala/org/http4s/util/CaseInsensitiveString.scala

```diff
@@ -1,13 +1,31 @@
 package org.http4s.util
 
 import java.util.Locale
+import scala.collection.generic.CanBuildFrom
+import scala.collection.{AbstractSeq, IndexedSeqOptimized}
+import scala.collection.immutable.{IndexedSeq, StringLike, WrappedString}
+import scala.collection.mutable.{Builder, StringBuilder}
+import scala.reflect.ClassTag
+import scalaz.{Equal, Monoid, Order, Ordering, Show}
 
-sealed class CaseInsensitiveString private (val value: String) extends CharSequence {
+/**
+ * A String wrapper such that two strings `x` and `y` are equal if
+ * `x.value.equalsIgnoreCase(y.value)`
+ */
+sealed class CaseInsensitiveString private (val value: String)
+    extends IndexedSeq[Char]
+    with IndexedSeqOptimized[Char, CaseInsensitiveString]
+    with Ordered[CaseInsensitiveString] {
   import CaseInsensitiveString._
 
-  private lazy val folded = value.toLowerCase(Locale.ROOT)
+  private var hash = 0
 
-  override def hashCode(): Int = folded.##
+  override def hashCode(): Int = {
+    if (hash == 0 && value.length > 0) {
+      hash = value.toLowerCase(Locale.ROOT).hashCode
```

Comment:
We'd save the array copy, but pay for bounds checking each value in charAt.  Might still be a win?